### PR TITLE
Contributing guide Improvement  Fix #331

### LINKED ISSUE #331: Contributing guide Improvement
Motivation
----------
From the current contributing guide it's not clear that pull requests depending on their types should go against certain branches.  
In addition, usual and popular pull request tools like the github interface and the hub wrapper are badly covered.

Modification
------------

* Few corrections for the format of the pull requests
* Remind the logic for patch and feature pull requests 
* Mention the github interface and the equivalent workflow with hub

### PR DESCRIPTION
**Pull request by vingar**

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -46,8 +46,34 @@ Every issue will get a **unique issue number**.
 
 **Step 2**: Create a local branch that corresponds to the issue. There are utility scripts to help you with this::
 
-  $ ./tools/create-patch-branch <unique issue number> '<component> #<issue number>: <short_change_message>'
-  $ ./tools/create-feature-branch <unique issue number> '<component> #<issue number>: <short_change_message>'
+  $ ./tools/create-patch-branch <unique issue number> '<short_change_message> #<issue number>: '
```
Comment (mlassnig) [inline CONTRIBUTING.rst]: i'd like to have the component in there, separated by a colon. what do others think?

"clients: my change #123"

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -46,8 +46,34 @@ Every issue will get a **unique issue number**.
...
+  $ ./tools/create-feature-branch <unique issue number> '<short_change_message> #<issue number>'
+
+The message can be possibly with
```
Comment (mlassnig) [inline CONTRIBUTING.rst]: Rewrite proposal:

"If you add a github-recognised [keyword](https://help.github.com/articles/closing-issues-using-keywords/) then the associated issue can be closed automatically once the pull request is merged."

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -46,8 +46,34 @@ Every issue will get a **unique issue number**.
...
+a [keyword](https://help.github.com/articles/closing-issues-using-keywords/) to close issue
+automatically when the pull request is merged. e.g.:
+
+    <short_change_message> Fix #<issue number>
+
+If you use different tools to create pull requests like the `github interface <https://help.github.com/articles/creating-a-pull-request/`_
+or the `hub the git command-line wrapper <https://hub.github.com`_, the following logic
+should be applied::
```
Comment (mlassnig) [inline CONTRIBUTING.rst]: s/should/must/

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -46,8 +46,34 @@ Every issue will get a **unique issue number**.
...
+
+* If the contribution is a **patch**, the pull request should go against the **next** and
```
Comment (mlassnig) [inline CONTRIBUTING.rst]: s/should/must/

File: CONTRIBUTING.rst [hunk 5]
```diff
@@ -46,8 +46,34 @@ Every issue will get a **unique issue number**.
...
+the **master** branch.
+
+* If the contribution is a new **feature**, the pull request should go against the **next**
```
Comment (mlassnig) [inline CONTRIBUTING.rst]: s/should/must/

File: PULL_REQUEST_TEMPLATE.md [hunk 6]
```diff
@@ -3,9 +3,9 @@ Pull request title
 
 The format of the Pull request title must be:
 
-    <component>: <short_change_message> #<issue number>
+    <short_change_message> #<issue number>
```
Comment (mlassnig) [inline PULL_REQUEST_TEMPLATE.md]: again, i like to have the component in there. the labels for components only show up on github, but never in the gitlog

File: CONTRIBUTING.rst (hunk 1 above)
Comment (vingar) [reply CONTRIBUTING.rst]: The component is redundant with the label and generate some repetition in the final release notes.

Comment (bari12) [reply CONTRIBUTING.rst]: For me the important part is the commit message, there the component should be mentioned and the guide already specifies it like this. In the branch name just `$ ./tools/create-patch-branch <unique issue number> '<short_change_message>` is sufficient. There is no point to have the #issue number at the end, it is already mentioned before and the # is filtered anyway as it is a branch name.

File: CONTRIBUTING.rst [hunk 7]
```diff
@@ -42,23 +42,48 @@ Contributing
...
-  $ ./tools/create-feature-branch <unique issue number> '<component> #<issue number>: <short_change_message>'
+  $ ./tools/create-patch-branch <unique issue number> '<short_change_message> #<issue number>: '
+  $ ./tools/create-feature-branch <unique issue number> '<short_change_message> #<issue number>'
+
+If you add a `github-recognised keyword <https://help.github.com/articles/closing-issues-using-keywords/>`_ then
+the associated issue can be closed automatically once the pull request is merged, e.g.::
+
+<short_change_message> Fix #<issue number>
+
+If you use different tools to create pull requests like the `github interface <https://help.github.com/articles/creating-a-pull-request/>`_
```
Comment (bari12) [inline CONTRIBUTING.rst]: This part should be in Step 4, not step 2 (creating the branch)

File: CONTRIBUTING.rst [hunk 8]
```diff
@@ -42,23 +42,48 @@ Contributing
...
+  $  git pull-request  -m  '<short_change_message> #<issue number>' -b next
+
+and for a new **feature**::
+
+  $  git pull-request  -m  '<short_change_message> #<issue number>' -b next
 
 **Step 3**: Commit your change. The format of the commit message must be::
 
-<component>: <change_message> #<issue number>
+<change_message> #<issue number>
```
Comment (bari12) [inline CONTRIBUTING.rst]: For the commit message I think it is fine to have the component.

File: PULL_REQUEST_TEMPLATE.md (hunk 6 above)
Comment (bari12) [reply PULL_REQUEST_TEMPLATE.md]: What the tools/submit-pull-request does is it replicates the commit message as the body of the PR, so it also includes the <component>.

File: CONTRIBUTING.rst (hunk 7 above)
Comment (vingar) [reply CONTRIBUTING.rst]: I did a confusion.  I'm wondering if an option  --fixed should not be added to /tools/submit-pull-request to add the github keywords to the pull request title.

Comment (bari12) [reply CONTRIBUTING.rst]: For me personally, I just add the Keyword in the commit message, if the ticket really should be closed. I don't see a big advantage to have --fixed in the tool, as the commit message still has to be written by hand, but it's also fine for me to add it. It's just a search for the # and add a Fix before 😄

File: PULL_REQUEST_TEMPLATE.md (hunk 6 above)
Comment (vingar) [reply PULL_REQUEST_TEMPLATE.md]: are the duplicated component names removed from the release notes then ? If yes, we can let the component here

Comment (bari12) [reply PULL_REQUEST_TEMPLATE.md]: The release notes are made from the issues, not from the commit messages. So in the issues they should not be mentioned, as we have the labels there (Also for search-ability). In the commit messages it is still nice to have them I think, as one might want to go through git log and easily see which components are involved.

Comment (vingar) [reply PULL_REQUEST_TEMPLATE.md]: fixed

File: CONTRIBUTING.rst (hunk 7 above)
Comment (vingar) [reply CONTRIBUTING.rst]: it is not clear than the commit message is propagated in the pull request title in the text. added it.

File: CONTRIBUTING.rst [hunk 9]
```diff
@@ -42,23 +42,55 @@ Contributing
 
 **Step 1**: Create an `issue <https://github.com/rucio/rucio/issues/new>`_ with the description
 of the contribution (motivation, modification and expected results).
-Every issue will get a **unique issue number**.
+Every issue will get a **unique issue number**. Valid component names are
+listed in the `label list <https://github.com/rucio/rucio/labels>`_.
```
Comment (bari12) [inline CONTRIBUTING.rst]: Not sure if the component name line is really necessary here. We have it below anyway where it is really needed.

Comment (vingar) [reply CONTRIBUTING.rst]: as a label

**Comment by mlassnig** [comment]
I've lost track, I leave it to @bari12 to approve ;-)

File: CONTRIBUTING.rst (hunk 9 above)
Comment (bari12) [reply CONTRIBUTING.rst]: I mean the sentence: "Valid component names..." is not really necessary here I feel.

File: CONTRIBUTING.rst [hunk 10]
```diff
@@ -42,23 +42,55 @@ Contributing
...
 
 **Step 4**: Push the commit to your forked repository and create the pull request(s). There is a helper script to assist you::
 
   $ ./tools/submit-pull-request
 
+The helper script will propagate the commit message as the pull request title.
+
+If you use different tools to create pull requests like the `github interface <https://help.github.com/articles/creating-a-pull-request/>`_
+or the git command-line wrapper `hub <https://hub.github.com>`_, the following logic must be applied:
+ * If the contribution is a **patch**, the pull request must go against the **next** and the **master** branch.
```
Comment (bari12) [inline CONTRIBUTING.rst]: "must go" might be a bit confusing for some readers. I would write:

> If the contribution is a patch, two pull requests must be created, one for the next and one for the master branch. 

Same for the feature.

File: CONTRIBUTING.rst [hunk 11]
```diff
@@ -42,23 +42,55 @@ Contributing
...
+ * If the contribution is a new **feature**, the pull request must go against the **next** branch.
+
+The format of the pull request title must be:
+
+    <component>: <short_change_message> #<issue number>
+
+If you add a `github-recognised keyword <https://help.github.com/articles/closing-issues-using-keywords/>`_ then
+the associated issue can be closed automatically once the pull request is merged, e.g.::
+
+<short_change_message> Fix #<issue number>
```
Comment (bari12) [inline CONTRIBUTING.rst]: For the sake of completeness:
`<component>: <short_change_message> Fix #<issue number>`

File: CONTRIBUTING.rst [hunk 12]
```diff
@@ -42,23 +42,55 @@ Contributing
...
+
+For example, with `hub the git command-line wrapper <https://hub.github.com>`_ it would be for
```
Comment (bari12) [inline CONTRIBUTING.rst]: > For example, with hub the git command-line wrapper the commands for a patch are:

File: CONTRIBUTING.rst (hunk 9 above)
Comment (vingar) [reply CONTRIBUTING.rst]: I dropped it. Is it possible for a non contributor to assign labels to an issue ?

Comment (bari12) [reply CONTRIBUTING.rst]: I don't think so.

Comment (vingar) [reply CONTRIBUTING.rst]: I removed it already. PR can be finally approved :)

**Review by bari12 — APPROVED** [review]

**Pull request merged**